src/analysis/check_checker/mod.rs

```diff
@@ -0,0 +1,2097 @@
+                    spans: vec![span],
+                    suggestion: None,
+                };
+                self.diagnostics.push(diagnostic);
+            }
+        }
+    }
+
+    // Filter any taint sources used in this expression
+    fn filter_taint(&mut self, expr: &SymbolicExpression) {
+        let node = Node::Expr(expr.id);
+        // Remove this node from the set of tainted nodes
+        if let Some(removed_node) = self.tainted_nodes.remove(&node) {
+            // Remove its sources of taint
+            for source_node in &removed_node.sources {
+                let source = self.taint_sources.remove(&source_node).unwrap();
+                self.tainted_nodes.remove(&source_node);
+                // Remove each taint source from its children
+                for child in &source.children {
+                    if let Some(mut child_node) = self.tainted_nodes.remove(child) {
+                        child_node.sources.remove(&source_node);
+                        // If the child is still tainted (by another source), add it back to the set
+                        if child_node.sources.len() > 0 {
+                            self.tainted_nodes.insert(child.clone(), child_node);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Check if the expression is annotated with `allow(unchecked_data)`
+    fn allow_unchecked(&mut self, span: &Span) -> bool {
+        // Since we are traversing in file order, we never need to check an
+        // annotation with a lower line number than the last check.
+        while self.annotation_index < self.annotations.len() {
+            let annotation = &self.annotations[self.annotation_index];
+            if annotation.span.start_line == (span.start_line - 1) {
+                return true;
+            } else if annotation.span.start_line >= span.start_line {
+                // The annotations are ordered by span, so if we have passed
+                // the target line, return.
+                return false;
+            }
+            self.annotation_index = self.annotation_index + 1;
+        }
+        false
+    }
+}
+
+impl<'a> ASTVisitor<'a> for CheckChecker<'a, '_> {
+    fn traverse_expr(&mut self, expr: &'a SymbolicExpression) -> bool {
+        // If this expression is annotated to allow unchecked data, no need to
+        // traverse it.
+        if self.allow_unchecked(&expr.span) {
+            return true;
+        }
+
+        match &expr.expr {
+            AtomValue(value) => self.visit_atom_value(expr, value),
+            Atom(name) => self.visit_atom(expr, name),
+            List(exprs) => self.traverse_list(expr, &exprs),
+            LiteralValue(value) => self.visit_literal_value(expr, value),
+            Field(field) => self.visit_field(expr, field),
+            TraitReference(name, trait_def) => self.visit_trait_reference(expr, name, trait_def),
+        }
+    }
+
+    fn traverse_define_public(
```

Comment:
Unless there is some sort of cascade already in place, I think we should treat `traverse_define_private` in a similar fashion: a public-function could delegate its writing operations to a private-function.

Reply:
When the public function is called, all of the parameters are added to the taint sources. Then, all expressions within that public function are checked. That means if a private function is called, it will already be properly checked. I'll add a unit test to ensure this.

Reply:
Oh, but we do need to transfer the taint onto the parameters, similar to the way `let` is handled. I will add that.

Reply:
This is actually a bit more complex than I initially thought, to get the most developer-friendly behavior. Here is what I am thinking. @lgalabru, let me know if you have any thoughts.
1. Require all checking to be done in the public function, i.e. only checked values can be passed to a call to a private function - this keeps the checking close to the source, which is desirable for readability
2. Add an annotation to allow a specific parameter of a private function to be unchecked - this allows for the use case of a private function which is a filter for unchecked data (dirty in, clean out)
3. If unchecked data is allowed (via 2), the return value must be a checked value

Reply:
604066c adds 1 and 2, except that the annotation applies to all parameters for now. Later, we can enhance this to specify which parameters are allowed to be unchecked. It also does not currently check the return value.

Reply:
The return value checking requires some more significant changes, so I will include that in a separate commit.